src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -0,0 +1,548 @@
+use smallvec::SmallVec;
```

Comment:
Please start the file with a doc comment summarizing the key things implemented here